File: 4817-2.jl
abstract type AbstractPoint end

mutable struct MyPoint <: AbstractPoint
    x::Float64
    y::Float64
end

function Base.show(io::IO, p::MyPoint)
    println(io, "($(p.x), $(p.y))")
end

function distance_to(p1::MyPoint, p2::MyPoint)
    return sqrt((p1.x - p2.x)^2 + (p1.y - p2.y)^2)
end

mutable struct MyGroup <: AbstractPoint
    name::String
    pivot::MyPoint
    members::Vector{MyPoint}
    old_members::Vector{MyPoint}
end

function Base.show(io::IO, g::MyGroup)
    println(io, "【組別 $(g.name)】")
    println(io, "中心點: $(g.pivot)")
    println(io, "成員: ", join(g.members, ", "))
end

function no_member_update(g::MyGroup)::Bool
    return g.old_members == g.members
end

function update_pivot(g::MyGroup)
    if isempty(g.members)
        return
    end
    sum_x = sum(p.x for p in g.members)
    sum_y = sum(p.y for p in g.members)
    n = length(g.members)
    g.pivot.x = sum_x / n
    g.pivot.y = sum_y / n

    g.old_members = copy(g.members)
    empty!(g.members)
end

mutable struct MySample <: AbstractPoint
    point::MyPoint
    group::Union{MyGroup, Nothing}
end

function Base.show(io::IO, s::MySample)
    println(io, "$(s.point)")
end

function set_group!(s::MySample, groups::Vector{MyGroup})
    distances = [distance_to(s.point, group.pivot) for group in groups]
    min_dist_group = argmin(distances)
    s.group = groups[min_dist_group]
    push!(groups[min_dist_group].members, s.point)
end

# 主函數
function main()
    # 樣本點
    samples_data = [
        (2.0, 5.0), (3.0, 2.0), (3.0, 3.0), (3.0, 4.0), (4.0, 3.0),
        (4.0, 4.0), (6.0, 3.0), (6.0, 4.0), (6.0, 6.0), (7.0, 2.0),
        (7.0, 5.0), (7.0, 6.0), (7.0, 7.0), (8.0, 6.0), (8.0, 7.0)
    ]

    # 分組點
    group_data = [
        ("1", 2.0, 2.0), ("2", 4.0, 6.0),
        ("3", 6.0, 5.0), ("4", 8.0, 8.0)
    ]

    # 轉換樣本點
    samples = [MySample(MyPoint(x, y), nothing) for (x, y) in samples_data]

    # 轉換分組點
    groups = [MyGroup(string(id), MyPoint(x, y), [], []) for (id, x, y) in group_data]

    # 搜索並更新直到沒有成員更改
    iteration = 0
    while true
        iteration += 1
        println("\n第 $iteration 次迭代")

        # 搜索
        for sample in samples
            set_group!(sample, groups)
        end

        # 平均距離
        mean_distance = sum(distance_to(sample.point, sample.group.pivot) for sample in samples) / length(samples)
        println("\nK-means 平均距離: $mean_distance")

        # 檢查聚類是否完成
        if all(group -> no_member_update(group), groups)
            break
        end

        # 更新
        for group in groups
            update_pivot(group)
        end
    end
end

main()

File: 4817-2.py
class Point:
    def __init__(self, x, y) -> None:
        self.x = x
        self.y = y
    
    def __str__(self) -> str:
        return f"({self.x}, {self.y})"
    
    def Euclidean_Distance(self, other_point) -> float:
        return ((self.x - other_point.x) ** 2 + (self.y - other_point.y) ** 2) ** 0.5

class Test:
    def __init__(self, name, x, y) -> None:
        self.name = name
        self.point = Point(x, y)
        self.value = []
        self.oldvalue = None
        
    def __str__(self) -> str:
        value_str = ", ".join(str(value) for value in self.value)
        return f"【Test {self.name}】\npoint: {self.point}\nvalue: {value_str}"
    
    def no_value_update(self) -> bool:
        if self.value == []:
            return True
        return self.value == self.oldvalue
    
    def update_point(self):
        if self.value:
            total_x = sum(value.x for value in self.value)
            total_y = sum(value.y for value in self.value)
            new_x = total_x / len(self.value)
            new_y = total_y / len(self.value)
            self.point = Point(new_x, new_y)
            self.oldvalue = list(self.value)
            self.value.clear()

class Sample(Point):
    def set_test(self, tests) -> None:
        distances = {test.name: self.Euclidean_Distance(test.point) for test in tests}
        min_test_name = min(distances, key=distances.get)
        min_test = next(test for test in tests if test.name == min_test_name)
        min_test.value.append(self)
        self.group = min_test  

    def distance_to_group_point(self):
        return self.Euclidean_Distance(self.group.point)

if __name__ == "__main__":    
    with open("C:/Users/bluew/OneDrive/桌面/HW1 1114817/sample points.csv", "r") as fin:
        samples_data = [line.strip().split(",") for line in fin]
        samples = [Sample(int(x), int(y)) for x, y in samples_data]
    
    with open("C:/Users/bluew/OneDrive/桌面/HW1 1114817/test points.csv", "r") as fin:
        tests_data = [line.strip().split(",") for line in fin]
        tests = [Test(name, int(x), int(y)) for name, x, y in tests_data]
        

        
    completed_iterations = 0
    max_iterations = 5
    mean_distances = []  
    for iteration in range(1, max_iterations + 1):
        print(f"\n{'Iteration ' + str(iteration)}")

        for sample in samples:
            sample.set_test(tests)

        print("\n\n".join(str(test) for test in tests))

        if all(test.no_value_update() for test in tests):
            break

        for test in tests:
            test.update_point()

       
        mean_distance = sum([sample.distance_to_group_point() for sample in samples]) / len(samples)
        mean_distances.append(mean_distance)
        print(f"\nMean distance after Iteration {iteration}: {mean_distance}")

    if completed_iterations == max_iterations:
        print("Clustering did not converge within the maximum number of iterations.")
    
    
    print("\nAll mean distances:", mean_distances)


import matplotlib.pyplot as plt


for sample in samples:
    plt.scatter(sample.x, sample.y, color='blue', marker='x', label='Sample')


for test in tests:
    plt.scatter(test.point.x, test.point.y, color='green', marker='*', s=200, label=f'Final Test {test.name} Center', zorder=2)


for test in tests:
    plt.scatter(test.point.x, test.point.y, color='red', marker='o', label=f'Initial Test {test.name}', zorder=3)

plt.xlabel('X')
plt.ylabel('Y')
plt.title('K-means Clustering')
plt.legend()
plt.grid(True)
plt.show()






fig, axs = plt.subplots(1, 3, figsize=(15, 5))


for sample in samples:
    axs[0].scatter(sample.x, sample.y, color='blue', marker='x', label='Sample')


for test in tests:
    axs[1].scatter(test.point.x, test.point.y, color='green', marker='*', s=200, label=f'Test {test.name} Center')


for test in tests:
    axs[2].scatter(test.point.x, test.point.y, color='red', marker='o', label=f'Test {test.name}')


axs[0].set_title('Sample Points')
axs[0].set_xlabel('X')
axs[0].set_ylabel('Y')

axs[1].set_title('Final Test Centers')
axs[1].set_xlabel('X')
axs[1].set_ylabel('Y')

axs[2].set_title('Initial Test Centers')
axs[2].set_xlabel('X')
axs[2].set_ylabel('Y')


axs[0].legend()
axs[1].legend()
axs[2].legend()


plt.tight_layout()
plt.show()

